src/cargo/core/workspace.rs

```diff
@@ -1702,10 +1702,18 @@ fn find_workspace_root_with_loader(
     mut loader: impl FnMut(&Path) -> CargoResult<Option<PathBuf>>,
 ) -> CargoResult<Option<PathBuf>> {
     // Check if there are any workspace roots that have already been found that would work
-    for (ws_root, ws_root_config) in config.ws_roots.borrow().iter() {
-        if manifest_path.starts_with(ws_root) && !ws_root_config.is_excluded(manifest_path) {
-            // Add `Cargo.toml` since ws_root is the root and not the file
-            return Ok(Some(ws_root.join("Cargo.toml").clone()));
+    {
+        let mut parent = manifest_path.parent();
+        let roots = config.ws_roots.borrow();
+        while let Some(current) = parent {
```

Comment:
I worry that changes here walk the file path similar to [`find_root_iter`](https://github.com/rust-lang/cargo/blob/b1dd22e668af5279e13a071ad4b17435bd6bfa4c/src/cargo/core/workspace.rs#L1731) which is used below this when [calling the loader](https://github.com/rust-lang/cargo/blob/b1dd22e668af5279e13a071ad4b17435bd6bfa4c/src/cargo/core/workspace.rs#L1712-L1717). Maybe there is some way to get it to work down there but I am not certain. Changing from a `HashMap` to a `BTreeMap` may be a solution to this since it is ordered.

One of the maintainers may have a better idea of what to do

Reply:
Using a `BTreeMap` [here](https://github.com/rust-lang/cargo/blob/b1dd22e668af5279e13a071ad4b17435bd6bfa4c/src/cargo/util/config/mod.rs#L206) and using `.rev()` [here](https://github.com/rust-lang/cargo/blob/b1dd22e668af5279e13a071ad4b17435bd6bfa4c/src/cargo/core/workspace.rs#L1705). Appears to fix the problem since `BTreeMap` is ordered ([example here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aa4ce5ad5133701942ace1948d8a060b)). I haven't been able to test if this solves the problem with nested workspaces but it should since it "walks" the file system.

Reply:
The file path walking shouldn't matter since this code doesn't do insertions, so the logic in `find_root_iter` will still apply to it. If I was to change the data structure instead, I think I would actually just use a simple `Vec` as honestly it feels like using either a HashMap or BTreeMap is overkill here given how few workspaces are actually ever going to be present.

Reply:
I missed that the logic for `find_root_iter` would apply here, that's my bad. Since that is the case, `BTreeMap` doesn't have any inherent advantages so you can disregard my suggestion if you wish.  The only tweak I would suggest is to use `.ancestors()` as it would be clearer what is going on.